### PR TITLE
Implement trace looping.

### PR DESCRIPTION
**Pull request by ptersilie**
Trace looping PR

This PR does several things. First in bab39a1e787cc73f484e9396ce4ddccbb4604212 we simplify traces by removing an unnecessary guard failure around the control point, while we are here, we insert an additional branch `loopentry`, which we'll use soon.

Where before we had:

```llvm
%5 = getelementptr %YkCtrlPointVars, ptr %0, i32 0, i32 0
%6 = load ptr, ptr %5, align 8
%7 = icmp eq null, null
br i1 %7, label %guardsuccess, label %guardfail

guardfail:
...

guardsuccess:
store ptr %6, ptr @shadowstack_0, align 8
...
```

We now get:

```llvm
entry:
%5 = getelementptr %YkCtrlPointVars, ptr %0, i32 0, i32 0
%6 = load ptr, ptr %5, align 8
br label %2

loopentry:
store ptr %6, ptr @shadowstack_0, align 8
...
```

In 36a65c103cb4051b5d5dedcd23ef2f0f96440124 we implement trace looping, which is as simple as replacing the return at the end of the trace with a `br` instruction to the `loopentry` block. We don't need to worry about insert PHI nodes since all the important variables are on the stack/shadowstack.

First performance tests have resulted in a slight (~15%) slowdown, due to traces now taking longer to compile. Presumably, this is because having a loop allows LLVM to apply more optimisations.

### CODE REVIEW

**Comment by vext01** [comment]
If you discount the trace compilation time, is there a speedup?

**Comment by ptersilie** [comment]
> If you discount the trace compilation time, is there a speedup?

No, it looked pretty similar, but I didn't really look further into it.

File: tests/c/debug_simple_interp_loop2.c [hunk 1]
```diff
@@ -42,7 +42,7 @@
 //       %{{fptr2}} = getelementptr %YkCtrlPointVars, ptr %0, i32 0, i32 0...
 //       store...
 //       ...
-//       ret ptr null
+//       br...
```
Comment (vext01) [inline tests/c/debug_simple_interp_loop2.c:45]: why don't we match the label name?

**Comment by vext01** [comment]
LGTM. I expected it to be harder!

**Comment by vext01** [comment]
So it's just "guard failure" -> "guard" where I pointed it out earlier, and a query about the tests.

File: tests/c/debug_simple_interp_loop2.c (hunk 1 above)
Comment (ptersilie) [reply tests/c/debug_simple_interp_loop2.c:45]: We might change the label at some point? There is a test that checks this while being agnostic about the chosen block name: [tests/c/simple.c](https://github.com/ykjit/yk/pull/736/commits/36a65c103cb4051b5d5dedcd23ef2f0f96440124#diff-30045133b33c3db0b2bd22080a981b5cd9d1c5d79f8b1b3cf50c2bc2cbdec05f)

**Comment by ptersilie** [comment]
> Do we have any tests which have post opt IR?

Not that I am aware of? @vext01?

**Comment by ltratt** [comment]
>>    Do we have any tests which have post opt IR?
>Not that I am aware of! @vext01?

I think it would be good to have at least 1, maybe even 2 or 3. Preferably starting with something ludicrously small where pre-opt probably can't optimise it. Such an example is a bit harder to create than it first appears, though.

**Comment by vext01** [comment]
5 tests use `jit-post-opt`, and I've added more in my promote branch (pre-opt has too much guff around which makes it hard to test).

**Comment by ltratt** [comment]
> 5 tests use jit-post-opt, and I've added more in my promote branch (pre-opt has too much guff around which makes it hard to test).

OK, maybe we should wait for the promote PR then.

**Comment by ptersilie** [comment]
I'm happy to wait, but at the same time, is any of this going to affect the changes here? We need trace looping one way or another, don't we?

File: tests/c/debug_simple_interp_loop2.c (hunk 1 above)
Comment (ltratt) [reply tests/c/debug_simple_interp_loop2.c:45]: Just this one and then I think we'll be ready to go.

**Comment by vext01** [comment]
Last commit LGTM. Now just "guard failure" -> "guard".

File: tests/c/debug_simple_interp_loop2.c (hunk 1 above)
Comment (ltratt) [reply tests/c/debug_simple_interp_loop2.c:45]: I more meant: shouldn't we do it everywhere?

**Comment by ptersilie** [comment]
> Last commit LGTM. Now just "guard failure" -> "guard".

I've commented on this already: https://github.com/ykjit/yk/commit/bab39a1e787cc73f484e9396ce4ddccbb4604212#r117269657

**Comment by ltratt** [comment]
> I've commented on this already: https://github.com/ykjit/yk/commit/bab39a1e787cc73f484e9396ce4ddccbb4604212#r117269657

Agreed. So it's just the label issue (or not...) and we're done.

File: tests/c/debug_simple_interp_loop2.c (hunk 1 above)
Comment (ptersilie) [reply tests/c/debug_simple_interp_loop2.c:45]: I just did the minimal change to get the test working and am leaning towards one place is enough to check this. But it's not a hill I'm going to die on, if you think it's useful.

Comment (ltratt) [reply tests/c/debug_simple_interp_loop2.c:45]: Fair enough.

**Comment by ltratt** [comment]
Please squash and rebase.

**Comment by ptersilie** [comment]
Oh, need to rebase too.

**Comment by ltratt** [comment]
Yes, please force push.

**Comment by ptersilie** [comment]
Rebased.

Once this change is in, we can tear out some of the deoptimisation logic (in ykllvm), since the control point will now always return a pointer (whereas before it could return null if the trace ran succesfully).

**Comment by ptersilie** [comment]
Sorry, formatting failed. `cargo xtask fmt` doesn't work on my machine at the moment, and I thought I could get away with it. Ok to squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.

**Pull request merged**